scripts/write-decode-map.ts

```diff
@@ -12,15 +12,22 @@ function convertMapToBinaryTrie(
     legacy: Record<string, string>
 ) {
     const encoded = encodeTrie(getTrie(map, legacy));
+    const hex = encoded.map((v) => v.toString(36)).join(",");
 
     // Write the encoded trie to disk
     fs.writeFileSync(
         `${__dirname}/../src/generated/decode-data-${name}.ts`,
         `// Generated using scripts/write-decode-map.ts
+/* eslint-disable */
 // prettier-ignore
-export default new Uint16Array([${encoded
-            .map((val) => val.toString(10))
-            .join(",")}]);
+export default /* #__PURE__ */ (function () {
+    const hex = "${hex}".split(',');
+    const arr = new Uint16Array(${encoded.length});
+    for (let i = 0; i < arr.length; i++) {
+        arr[i] = parseInt(hex[i], 36);
+    }
+    return arr;
+})();
 `
     );
```

Comment:
I've played with making this base64, for an even compacter encoding. Curious what you think:

```js
const encoded = new Uint16Array(encodeTrie(getTrie(map, legacy)));
// eslint-disable-next-line node/no-unsupported-features/node-builtins
const base64 = Buffer.from(encoded.buffer).toString("base64");

/*
    * We don't want to depend on Node's Buffer, so we use
    * the native `atob` function.
    *
    * Let's make sure the decoding works as expected.
    */
assert.deepEqual(
    new Uint16Array(
        new Uint8Array(
            atob(base64)
                .split("")
                .map((c) => c.charCodeAt(0))
        ).buffer
    ),
    encoded
);

// Write the encoded trie to disk
fs.writeFileSync(
    `${__dirname}/../src/generated/decode-data-${name}.ts`,
    `// Generated using scripts/write-decode-map.ts
// prettier-ignore
export default new Uint16Array(new Uint8Array(atob("${base64}").split('').map(c=>c.charCodeAt(0))).buffer);
`
);
```

Reply:
This is a smart way to do it, I hadn't thought of using new Uint8Array.
I took more detailed data and found that it greatly reduces the raw size, but increases the gzipped size.

|               | raw size<br>(w/o terser) | raw size<br>(w/ terser) | gzip size<br>(w/ terser) | brotli size<br>(w/ terser) |
|----------|-----------------------------|---------------------------|----------------------------|-----------------------------|
| master branch | 130261B | 98431B | 31629B | 25828B |
| base36 | 100718B (-22.7%) | 83939B (-14.7%) | 30569B (-3.4%) | 25256B (-2.2%) |
| base64 | 89979B (-31.0%) | 73438B (-25.4%) | 32148B (**+1.6%**) | 28135B (**+8.9%**) |

Reply:
Interesting — makes some sense given that patterns aren't as repetitive anymore, but I did expect this to be less of a factor. As yet another alternative, there is always storing the binary data directly in a string:

```js
const encoded = encodeTrie(getTrie(map, legacy));
const stringified = JSON.stringify(String.fromCharCode(...encoded));

// Write the encoded trie to disk
fs.writeFileSync(
    `${__dirname}/../src/generated/decode-data-${name}.ts`,
    `// Generated using scripts/write-decode-map.ts
// prettier-ignore
export default new Uint16Array(
  ${stringified}
    .split("")
    .map((c) => c.charCodeAt(0))
);
`
);
```

Reply:
This is not ideal, as it adds the trap door of character encodings, but for the sake of argument 🙂

Reply:
For context, this reduces the file size for `decode-data-html` by 58%.

Reply:
I avoided that way to avoid encoding problems at first thought. But it seems it reduces the size significantly.
To mitigate encoding problems, maybe escaping non-ascii non-control characters will work.
This increases raw size but terser will be unescape them unless [`format.ascii_only = true`](https://github.com/terser/terser#format-options) is set.
This escaping way reduces gzip/brotli file size than base36/base64.

|               | raw size<br>(w/o terser) | raw size<br>(w/ terser) | gzip size<br>(w/ terser) | brotli size<br>(w/ terser) |
|----------|-----------------------------|---------------------------|----------------------------|-----------------------------|
| master branch | 130261B | 98431B | 31629B | 25828B |
| base36 | 100718B (-22.7%) | 83939B (-14.7%) | 30569B (-3.4%) | 25256B (-2.2%) |
| base64 | 89979B (-31.0%) | 73438B (-25.4%) | 32148B (+1.6%) | 28135B (+8.9%) |
| binary string | 76693B (-41.1%) | 59324B (-39.7%) | 28783B (-9.0%) | 23668B (-8.4%) |
| binary string escaped<br> | 98479B (-24.4%) | 59324B (-39.7%) | 28783B (-9.0%) | 23668B (-8.4%) |
| binary string escaped<br> + terser ascii_only=true | 98479B (-24.4%) | 80257B (-18.5%) | 30335B (-4.1%) | 24507B (-5.1%) |

Reply:
I like it! Let's simplify the escaping logic to

```js
const stringified = JSON.stringify(String.fromCharCode(...encoded)).replace(
    /[^\x20-\x7e]/g,
    (c) => `\\u${c.charCodeAt(0).toString(16).padStart(4, "0")}`
);
```

This will save some bytes if a known escape (eg. `\t`) is found. The `prettier-ignore` is now unnecessary as well.

Reply:
Turns out we can use hex escapes to save a few more bytes (see https://mathiasbynens.be/notes/javascript-escapes):

```js
const stringified = JSON.stringify(String.fromCharCode(...encoded))
  .replace(
      /[^\0-\x7e]/g,
      (c) => `\\u${c.charCodeAt(0).toString(16).padStart(4, "0")}`
  )
  .replace(/\\u0000/g, "\\0")
  .replace(/\\u00(\d\d)/g, "\\x$1");
```

Reply:
Anyhow, this feels as far as this can be taken. Lmk what you think, and then let's get this merged!

Reply:
There is also https://github.com/mathiasbynens/jsesc, which could be added as a dev dependency and would remove the need for all of the `replace` logic.